### DISCUSSION TITLE
Attribute prioritization for launcher search

### DISCUSSION THREAD

**Discussion by majkinetor** [1Remote/1Remote#398] [Ideas]
It happens that I can't easily find something via launcher, if it wasn't executed since some time (or mostly, at all). I use main window then as search doesn't include keyword matcher options (whih are all enabled for me).

The problem is that some less important connection attributes hijack place of more important ones. If I have a keyword in a name, it should be a priority vs less important attributes such as username:

<details><summary>details</summary> 

![image](https://user-images.githubusercontent.com/85767/236670843-cc358b47-dbe6-49db-991c-a8fb086024d4.png)
</details>

So idea is to have some sort of algorythim to determine the order, such as.

Order of attributes:

1. usage frequency
1. name
2. hostname
3. username
4. any

Order of operations:

1. Fully equal
2. Starts with 
3. Contains with boundary ( regex \b )
4. Contains
6. Fuzzy

So in above case, if usage frequency is the same for all:

<details><summary>details</summary>

 ![image](https://user-images.githubusercontent.com/85767/236670990-1611c085-dfb7-41d9-968f-e9b34771f934.png)</details>

**Now, I guess this is NOT worth doing :) Just something on my mind so I thought I should write it up**

**Comment by VShawn** [comment]
Weighted text search requires more cpu resources, so it is necessary to carefully optimize the search algorithm.

**Reply by majkinetor** [reply]
Launcher search is already very slow without this, and it should be improved in the future. Its unresalisticlly slow given that I have around 50 connections. 

But, it can be cached, if nothing else. And not a big problem in this moment. Performance is something that can be done at the end. Most commercial tools are slow as hell too, even without fancy search.

**Reply by VShawn** [reply]
> `Launcher search is already very slow` 

The slow is due to slow UI updates: once a key is pressed, all of your 50 items need to rerender the highlighted strings, I'm not an expert in WPF frontend development, so I can hardly figured out the way to optimize them. 

BTW, it is cached in the Launcher now.

**Reply by majkinetor** [reply]
It seems main thread is blocked while you update the screen. This is obvious as if you type something and then delete slowly, it glitches while it paints. This is classical problem when less performant procedure blocks UI. Its typically solved in Win32 by updating UI in separate thread. This way main thread can terminate any running (and now outdated) thread and restart the process of painting.
